presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughOuterJoinRuleSet.java

```diff
@@ -0,0 +1,220 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.iterative.Rule.Context;
+import io.prestosql.sql.planner.optimizations.SymbolMapper;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.JoinNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.optimizations.QueryCardinalityUtil.extractCardinality;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.FULL;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.prestosql.sql.planner.plan.JoinNode.Type.RIGHT;
+import static io.prestosql.sql.planner.plan.Patterns.Join.type;
+import static io.prestosql.sql.planner.plan.Patterns.join;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+import static io.prestosql.sql.planner.plan.TopNNode.Step.PARTIAL;
+
+/**
+ * Transforms:
+ * <pre>
+ * - TopN
+ *    - [Project]
+ *       - Join
+ *          - left source
+ *          - right source
+ * </pre>
+ * Into:
+ * <pre>
+ * - TopN
+ *    - [Project]
+ *       - Join
+ *          - TopN (present if Join is left or outer, not already limited, and orderBy symbols come from left source)
+ *             - left source
+ *          - TopN (present if Join is right or outer, not already limited, and orderBy symbols come from right source)
+ *             - right source
+ * </pre>
+ */
+public final class PushTopNThroughOuterJoinRuleSet
+{
+    private PushTopNThroughOuterJoinRuleSet() {}
+
+    public static Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(withProject(), withoutProject());
+    }
+
+    @VisibleForTesting
+    static Rule<?> withProject()
+    {
+        return new PushTopNThroughOuterJoinWithProject();
+    }
+
+    @VisibleForTesting
+    static Rule<?> withoutProject()
+    {
+        return new PushTopNThroughOuterJoinWithoutProject();
+    }
+
+    private static class PushTopNThroughOuterJoinWithProject
```

Comment:
please extract `PushTopNThroughProject` rule as a separate top level rule.
With conditions that projection should not be on top of filter or table scan

Reply:
@sopel39 ,
- https://github.com/prestosql/presto/pull/220#discussion_r258433361 ?
- projection can be narrowing, so such a general rule would not be a good idea -- in the resulting plan, TopN would have higher mem requirements.
  Of course, we can have a rule with a condition that this is not a narrowing projection, but then, this would no longer deliver on the "push TopN through Outer Join" idea.

@kokosing what's your thought on that?

Reply:
> Patterns do not handle the "is not filter nor table scan", so I think we can end up having push-topn-through-x-y-or-z rules. And what we have here would just be one of them.

It does:

```
topN().with(source().matching(node -> !(node instanceof TableScanNode) && !(node instanceof FilterNode)));
```

Another option is to check that in the rule itself (by using lookup ; /), then it is not a "rule" anymore, but "PlanOptiizer".

>  projection can be narrowing

Obviously, projection can also produce symbols that are used for sorting. That makes difficult to push down sort operation through projection without introducing another projection underneath. So that would work only for partial identity (mapping only sort symbols) projections?

Also this could trigger an optimizer loop as we tend to pushdown narrowing projection everywhere if possible.

> * TopN would have higher mem requirements.

Do we care about it? I would expect TopN have a low limit for rows then mem requirements are not a big deal here.

> this would no longer deliver on the "push TopN through Outer Join" idea.

Correct, we would lose the context of this special case with generic rule. Theoretically, we lose opportunity to select a better plan, however I am not sure if it is a case here.

---
I would add such generic rule as separate PR and continue work on this PR.

Reply:
Pushing down TopN should almost always save CPU. In most cases that would be drastic CPU reduction (e.g: processing few rows compared to entire table). Therefore, I think we should push TopN as low as we can (e.g: below aggregation, windows) as a general principle (similar as for limit). Ideally, we could pushdown to connector itself (for JDBC this could make huge impact).

I think there should be more TopN pushdown rules (if there aren't yet):
* through aggregation
* through window
* through union/exchange

> in the resulting plan, TopN would have higher mem requirements.

I think TopN usually takes negligible memory. However, we could limit pushdown if N is small (e.g: 100, 1000).

Reply:
Even in case of pushing down through join you don't know wether it will help or not (join can also be narrowing). I think in such edge (almost negligible) cases we should use common sense (e.g: what would be typical use case, for what % of queries it will help)

Reply:
> I think we should push TopN as low as we can (e.g: below aggregation, windows)

Pushing TopN through aggregation changes the query semantic.